test/test_linalg.py

```diff
@@ -963,6 +963,147 @@ def get_value_type(dtype):
+            (10, ),
+            (10, 10),
+            (3, 4, 5),
+            (0, ),
+            (0, 10),
+            (10, 0),
+            (0, 0),
+        ]
+        for keepdim in [True, False]:
+            for input_size in input_sizes:
+                for ord in ord_settings:
+                    if 0 in input_size and ord in [inf, -inf]:
+                        continue
+                    dtypes = [torch.float, torch.double, torch.cfloat, torch.cdouble]
+                    for from_dtype, to_dtype in itertools.product(dtypes, dtypes):
+                        run_test_case(input_size, ord, keepdim, from_dtype, to_dtype)
+
+        # Make sure that setting dtype != out.dtype raises an error
+        dtype_pairs = [
+            (torch.float, torch.double),
+            (torch.double, torch.float),
+            (torch.cfloat, torch.cdouble),
+            (torch.cdouble, torch.cfloat),
+        ]
+        for keepdim in [True, False]:
+            for input_size in input_sizes:
+                for ord in ord_settings:
+                    for dtype, out_dtype in dtype_pairs:
+                        input = torch.rand(*input_size)
+                        result = torch.Tensor().to(out_dtype)
+                        with self.assertRaisesRegex(RuntimeError, r'provided dtype must match dtype of result'):
+                            torch.linalg.norm(input, ord=ord, keepdim=keepdim, dtype=dtype, out=result)
+
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_vector_norm(self, device, dtype):
+        # This test compares torch.linalg.vector_norm's output with
+        # torch.linalg.norm given a flattened tensor
+        ord_vector = [0, 0.5, 1, 2, 3, inf, -0.5, -1, -2, -3, -inf, None]
+        test_cases = [
+            # input size, ords that should cause error
+            ((10, ), []),
+            ((4, 5), []),
+            ((3, 4, 5), []),
+            ((3, 4, 5, 6, 7), []),
+
+            # Infinite orders cause an error for degenerate inputs because
+            # min/max cannot be called on degernate inputs
+            ((0, ), [inf, -inf]),
+            ((0, 10), [inf, -inf]),
+            ((10, 0), [inf, -inf]),
+            ((0, 0), [inf, -inf]),
+            ((10, 0, 10), [inf, -inf]),
+            ((0, 0, 0), [inf, -inf]),
+        ]
+
+        def run_test_case(input, ord, keepdim, should_error):
+            msg = f'input.size()={input.size()}, ord={ord}, keepdim={keepdim}, dtype={dtype}'
+            input_copy = input.clone().detach()
+            input_flat = input_copy.flatten(0, -1)
+            if should_error:
+                with self.assertRaises(RuntimeError):
+                    torch.linalg.norm(input_flat, ord, keepdim=keepdim)
+                with self.assertRaises(RuntimeError):
+                    torch.linalg.vector_norm(input, ord, keepdim=keepdim)
+            else:
+                result_preflattened = torch.linalg.norm(input_flat, ord, keepdim=keepdim)
```

Comment:
This is good but I suggest making a reference function, like vector_norm_reference(), with the same signature as vector norm except it uses this alternative (Python) implementation. Then these tests could just call that reference and compare the results.